File: backend/src/sandbox/layer_stack/squash.py
"""Checkpoint-based depth control for sandbox layer stacks."""

from __future__ import annotations

import os
import shutil
import uuid
from dataclasses import dataclass
from pathlib import Path

from sandbox.layer_stack.paths import (
    allocate_unique_layer_paths,
    fsync_path,
    resolve_storage_path,
)
from sandbox.layer_stack.manifest import LAYERS_DIR, STAGING_DIR, LayerRef, Manifest
from sandbox.layer_stack.view import MergedView


@dataclass(frozen=True)
class CheckpointSegment:
    layers: tuple[LayerRef, ...]

    def __post_init__(self) -> None:
        if len(self.layers) <= 1:
            raise ValueError("checkpoint segments must contain at least two layers")


_SquashPlanEntry = LayerRef | CheckpointSegment


@dataclass(frozen=True)
class SquashPlan:
    active_version: int
    active_layers: tuple[LayerRef, ...]
    entries: tuple[_SquashPlanEntry, ...]

    def __post_init__(self) -> None:
        if not self.active_layers:
            raise ValueError("active_layers must not be empty")
        if not self.entries:
            raise ValueError("entries must not be empty")
        if not self.checkpoint_segments:
            raise ValueError("squash plans must include at least one checkpoint segment")

    @property
    def checkpoint_segments(self) -> tuple[CheckpointSegment, ...]:
        return tuple(entry for entry in self.entries if isinstance(entry, CheckpointSegment))


class LayerCheckpointSquasher:
    """Plans runs between barrier layers and projects each run into a checkpoint layer."""

    def __init__(
        self,
        storage_root: str | Path,
    ) -> None:
        self._storage_root = Path(storage_root)
        self._view = MergedView(self._storage_root)

    def plan(
        self,
        active_manifest: Manifest,
        *,
        max_depth: int,
        barrier_layers: tuple[LayerRef, ...] = (),
        min_reduction: int = 1,
    ) -> SquashPlan | None:
        if max_depth <= 0:
            raise ValueError("max_depth must be positive")
        if min_reduction <= 0:
            raise ValueError("min_reduction must be positive")
        if active_manifest.depth <= max_depth:
            return None

        entries = _segment_around_barriers(active_manifest.layers, barrier_layers)
        if len(entries) >= active_manifest.depth:
            return None
        if active_manifest.depth - len(entries) < min_reduction:
            return None
        checkpoint_segments = tuple(
            entry for entry in entries if isinstance(entry, CheckpointSegment)
        )
        if len(entries) > max_depth and all(
            len(segment.layers) <= max_depth for segment in checkpoint_segments
        ):
            return None

        return SquashPlan(
            active_version=active_manifest.version,
            active_layers=active_manifest.layers,
            entries=entries,
        )

    def build_checkpoint(
        self,
        segment: CheckpointSegment,
        *,
        active_version: int,
    ) -> LayerRef:
        layer_id, staging_dir, layer_dir = self._allocate_checkpoint_paths(active_version + 1)
        segment_manifest = Manifest(
            version=active_version,
            layers=segment.layers,
        )
        try:
            self._view.materialize(staging_dir, segment_manifest)
            layer_dir.parent.mkdir(parents=True, exist_ok=True)
            os.replace(staging_dir, layer_dir)
        except Exception:
            shutil.rmtree(staging_dir, ignore_errors=True)
            raise
        return LayerRef(layer_id=layer_id, path=f"{LAYERS_DIR}/{layer_id}")

    def relabel_checkpoint(self, checkpoint: LayerRef, *, manifest_version: int) -> LayerRef:
        """Rename a prebuilt checkpoint to match the manifest that will publish it."""
        current_path = resolve_storage_path(self._storage_root, checkpoint.path)
        if not current_path.exists():
            raise FileNotFoundError(f"checkpoint layer is missing: {checkpoint.layer_id}")

        layer_id, _staging_dir, layer_dir = self._allocate_checkpoint_paths(
            manifest_version
        )
        os.replace(current_path, layer_dir)
        fsync_path(layer_dir.parent)
        return LayerRef(layer_id=layer_id, path=f"{LAYERS_DIR}/{layer_id}")

    def discard_checkpoint(self, checkpoint: LayerRef) -> None:
        layer_path = resolve_storage_path(self._storage_root, checkpoint.path)
        shutil.rmtree(layer_path, ignore_errors=True)

    def _allocate_checkpoint_paths(self, next_version: int) -> tuple[str, Path, Path]:
        return allocate_unique_layer_paths(
            storage_root=self._storage_root,
            layers_dir=LAYERS_DIR,
            staging_dir=STAGING_DIR,
            next_version=next_version,
            id_factory=_default_checkpoint_id,
        )


def _segment_around_barriers(
    layers: tuple[LayerRef, ...],
    barrier_layers: tuple[LayerRef, ...],
) -> tuple[_SquashPlanEntry, ...]:
    barriers = set(barrier_layers)
    entries: list[_SquashPlanEntry] = []
    run: list[LayerRef] = []

    def flush_run() -> None:
        if len(run) > 1:
            entries.append(CheckpointSegment(tuple(run)))
        elif run:
            entries.append(run[0])
        run.clear()

    for layer in layers:
        if layer in barriers:
            flush_run()
            entries.append(layer)
        else:
            run.append(layer)
    flush_run()
    return tuple(entries)


def manifest_prefix_before_plan(
    manifest: Manifest,
    plan: SquashPlan,
) -> tuple[LayerRef, ...] | None:
    planned_depth = len(plan.active_layers)
    if planned_depth > manifest.depth:
        return None
    if manifest.layers[-planned_depth:] != plan.active_layers:
        return None
    return manifest.layers[:-planned_depth]


def _default_checkpoint_id(next_version: int) -> str:
    return f"B{next_version:06d}-{uuid.uuid4().hex[:8]}"


__all__ = [
    "CheckpointSegment",
    "LayerCheckpointSquasher",
    "SquashPlan",
    "manifest_prefix_before_plan",
]

File: sandbox/crates/eos-layerstack/src/squash.rs
use std::path::PathBuf;

use crate::model::{LayerRef, Manifest, MANIFEST_SCHEMA_VERSION};

use crate::error::LayerStackError;
use crate::fs::{allocate_layer_dirs, check_layer_path, fsync_dir, resolve_layer_path};
use crate::{MergedView, LAYERS_DIR};

pub const CHECKPOINT_ID_PREFIX: char = 'B';

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CheckpointSegment {
    pub layers: Vec<LayerRef>,
}

impl CheckpointSegment {
    pub(crate) fn new(layers: Vec<LayerRef>) -> Result<Self, LayerStackError> {
        if layers.len() <= 1 {
            return Err(LayerStackError::InvalidSquashPlan(
                "checkpoint segments must contain at least two layers".to_owned(),
            ));
        }
        Ok(Self { layers })
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SquashPlanEntry {
    Keep(LayerRef),
    Segment(CheckpointSegment),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SquashPlan {
    pub active_version: i64,
    pub active_layers: Vec<LayerRef>,
    pub entries: Vec<SquashPlanEntry>,
}

impl SquashPlan {
    pub(crate) fn new(
        active_version: i64,
        active_layers: Vec<LayerRef>,
        entries: Vec<SquashPlanEntry>,
    ) -> Result<Self, LayerStackError> {
        if active_layers.is_empty() {
            return Err(LayerStackError::InvalidSquashPlan(
                "active_layers must not be empty".to_owned(),
            ));
        }
        if entries.is_empty() {
            return Err(LayerStackError::InvalidSquashPlan(
                "entries must not be empty".to_owned(),
            ));
        }
        if !entries
            .iter()
            .any(|e| matches!(e, SquashPlanEntry::Segment(_)))
        {
            return Err(LayerStackError::InvalidSquashPlan(
                "squash plans must include at least one checkpoint segment".to_owned(),
            ));
        }
        Ok(Self {
            active_version,
            active_layers,
            entries,
        })
    }

    #[must_use]
    pub fn checkpoint_segments(&self) -> Vec<&CheckpointSegment> {
        self.entries
            .iter()
            .filter_map(|e| match e {
                SquashPlanEntry::Segment(s) => Some(s),
                SquashPlanEntry::Keep(_) => None,
            })
            .collect()
    }
}

#[derive(Debug)]
pub struct LayerCheckpointSquasher {
    storage_root: PathBuf,
    view: MergedView,
}

impl LayerCheckpointSquasher {
    #[must_use]
    pub fn new(storage_root: PathBuf) -> Self {
        Self {
            view: MergedView::new(storage_root.clone()),
            storage_root,
        }
    }

    pub fn plan(
        &self,
        active_manifest: &Manifest,
        max_depth: usize,
        lease_head_layers: &[LayerRef],
        min_reduction: usize,
    ) -> Result<Option<SquashPlan>, LayerStackError> {
        if max_depth == 0 {
            return Err(LayerStackError::InvalidSquashPlan(
                "max_depth must be positive".to_owned(),
            ));
        }
        if min_reduction == 0 {
            return Err(LayerStackError::InvalidSquashPlan(
                "min_reduction must be positive".to_owned(),
            ));
        }
        if active_manifest.layers.len() <= max_depth {
            return Ok(None);
        }

        let entries = segment_around_lease_heads(&active_manifest.layers, lease_head_layers)?;
        if entries.len() >= active_manifest.layers.len() {
            return Ok(None);
        }
        if active_manifest.layers.len() - entries.len() < min_reduction {
            return Ok(None);
        }
        let plan = SquashPlan::new(
            active_manifest.version,
            active_manifest.layers.clone(),
            entries,
        )?;
        if plan.entries.len() > max_depth
            && plan
                .checkpoint_segments()
                .iter()
                .all(|segment| segment.layers.len() <= max_depth)
        {
            return Ok(None);
        }
        Ok(Some(plan))
    }

    pub(crate) fn build_checkpoint(
        &self,
        segment: &CheckpointSegment,
        active_version: i64,
    ) -> Result<LayerRef, LayerStackError> {
        let (layer_id, staging_dir, layer_dir) =
            allocate_layer_dirs(&self.storage_root, CHECKPOINT_ID_PREFIX, active_version + 1)?;
        let segment_manifest = Manifest::new(
            active_version,
            segment.layers.clone(),
            MANIFEST_SCHEMA_VERSION,
        )
        .map_err(LayerStackError::from)?;
        if let Err(err) = self.view.project(&staging_dir, &segment_manifest) {
            let _ = std::fs::remove_dir_all(&staging_dir);
            return Err(err);
        }
        if let Some(parent) = layer_dir.parent() {
            std::fs::create_dir_all(parent)?;
        }
        if let Err(err) = std::fs::rename(&staging_dir, &layer_dir) {
            let _ = std::fs::remove_dir_all(&staging_dir);
            return Err(err.into());
        }
        Ok(LayerRef {
            layer_id: layer_id.clone(),
            path: format!("{LAYERS_DIR}/{layer_id}"),
        })
    }

    pub(crate) fn relabel_checkpoint(
        &self,
        checkpoint: &LayerRef,
        manifest_version: i64,
    ) -> Result<LayerRef, LayerStackError> {
        let current = self.layer_path(checkpoint)?;
        if !current.exists() {
            return Err(LayerStackError::Storage(format!(
                "checkpoint layer is missing: {}",
                checkpoint.layer_id
            )));
        }
        let (layer_id, _staging_dir, layer_dir) =
            allocate_layer_dirs(&self.storage_root, CHECKPOINT_ID_PREFIX, manifest_version)?;
        if let Some(parent) = layer_dir.parent() {
            std::fs::create_dir_all(parent)?;
        }
        std::fs::rename(current, &layer_dir)?;
        if let Some(parent) = layer_dir.parent() {
            fsync_dir(parent)?;
        }
        Ok(LayerRef {
            layer_id: layer_id.clone(),
            path: format!("{LAYERS_DIR}/{layer_id}"),
        })
    }

    pub(crate) fn discard_checkpoint(&self, checkpoint: &LayerRef) -> Result<(), LayerStackError> {
        let path = self.layer_path(checkpoint)?;
        match std::fs::remove_dir_all(path) {
            Ok(()) => Ok(()),
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => Ok(()),
            Err(err) => Err(err.into()),
        }
    }

    fn layer_path(&self, layer: &LayerRef) -> Result<PathBuf, LayerStackError> {
        check_layer_path(&layer.path)?;
        Ok(resolve_layer_path(&self.storage_root, &layer.path))
    }
}

#[must_use]
pub(crate) fn manifest_prefix_before_plan<'m>(
    manifest: &'m Manifest,
    plan: &SquashPlan,
) -> Option<&'m [LayerRef]> {
    let planned_depth = plan.active_layers.len();
    if planned_depth > manifest.layers.len() {
        return None;
    }
    let split = manifest.layers.len() - planned_depth;
    if manifest.layers[split..] != plan.active_layers {
        return None;
    }
    Some(&manifest.layers[..split])
}

fn segment_around_lease_heads(
    layers: &[LayerRef],
    lease_head_layers: &[LayerRef],
) -> Result<Vec<SquashPlanEntry>, LayerStackError> {
    let mut entries = Vec::new();
    let mut run = Vec::new();
    for layer in layers {
        if lease_head_layers.contains(layer) {
            flush_run(&mut entries, &mut run)?;
            entries.push(SquashPlanEntry::Keep(layer.clone()));
        } else {
            run.push(layer.clone());
        }
    }
    flush_run(&mut entries, &mut run)?;
    Ok(entries)
}

fn flush_run(
    entries: &mut Vec<SquashPlanEntry>,
    run: &mut Vec<LayerRef>,
) -> Result<(), LayerStackError> {
    match run.len() {
        0 => {}
        1 => entries.push(SquashPlanEntry::Keep(run[0].clone())),
        _ => entries.push(SquashPlanEntry::Segment(CheckpointSegment::new(
            std::mem::take(run),
        )?)),
    }
    run.clear();
    Ok(())
}

#[cfg(test)]
#[path = "../tests/unit/squash.rs"]
mod tests;
